# Fetch - Data Analyst Take Home

## In this exercise you will:

**Demonstrate how you reason about data and how you communicate your understanding of a specific data set to others.**

## Requirements:

**The exercise must be committed to a git repository and the link must be provided to the recruiter. Emailed files will not be reviewed.**

## Tips:

- Communicate your thoughts in each section. Code should be commented. Findings and conclusions should be written down and shared.
- Share assumptions you are making.

## Exercise:

### First: Explore the Data

**Review the unstructured CSV files and answer the following questions with code that supports your conclusions:**

- Are there any data quality issues present?
- Are there any fields that are challenging to understand?

*We recommend using SQL or Python and data visualization to examine the data.*

In [1]:
#------------#
#  Packages  #
#------------#

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#----------------#
#  Load Dataset  #
#----------------#
products = pd.read_csv('PRODUCTS_TAKEHOME.csv', dtype={'BARCODE': str})
transactions = pd.read_csv('TRANSACTION_TAKEHOME.csv', dtype={'BARCODE': str})
users = pd.read_csv('USER_TAKEHOME.csv')

In [2]:
products.head()

CATEGORY_1              CATEGORY_2                   CATEGORY_3  \
0  Health & Wellness           Sexual Health  Conductivity Gels & Lotions   
1             Snacks           Puffed Snacks         Cheese Curls & Puffs   
2  Health & Wellness               Hair Care        Hair Care Accessories   
3  Health & Wellness               Oral Care                   Toothpaste   
4  Health & Wellness  Medicines & Treatments               Essential Oils   

  CATEGORY_4                                       MANUFACTURER  \
0        NaN                                                NaN   
1        NaN                                                NaN   
2        NaN                           PLACEHOLDER MANUFACTURER   
3        NaN                                  COLGATE-PALMOLIVE   
4        NaN  MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...   

             BRAND        BARCODE  
0              NaN   796494407820  
1              NaN   023278011028  
2          ELECSOP   461817824225  
3          COLGATE   035000466815  
4  MAPLE HOLISTICS  0806810850459

In [ ]:
# A function to check each csv to spot check at a glance for data quality issues
# Uses shape, columns, missing values, unique values, summary statistics, and duplicated rows

def data_quality_check(df):
    """
    This function returns print statements for quick glance to determine for data quality issues or fields that need to be cleaned
    
    Args:
    df: DataFrame
    
    Returns:
    print statements
    """
    print('The shape of the dataset is:', df.shape)
    print('The columns in the dataset are:', df.columns)
    print('The number of missing values in the dataset are:', df.isnull().sum())
    print('The number of unique values in the dataset are:', df.nunique())
    print('The summary statistics of the dataset are:', df.describe())
    print('The number of duplicated rows in the dataset are:', df.duplicated().sum())

# Check for data quality issues in the products dataset
data_quality_check(products)


The shape of the dataset is: (845552, 7)
The columns in the dataset are: Index(['CATEGORY_1', 'CATEGORY_2', 'CATEGORY_3', 'CATEGORY_4', 'MANUFACTURER',
       'BRAND', 'BARCODE'],
      dtype='object')
The number of missing values in the dataset are: CATEGORY_1         111
CATEGORY_2        1424
CATEGORY_3       60566
CATEGORY_4      778093
MANUFACTURER    226474
BRAND           226472
BARCODE           4025
dtype: int64
The number of unique values in the dataset are: CATEGORY_1          27
CATEGORY_2         121
CATEGORY_3         344
CATEGORY_4         127
MANUFACTURER      4354
BRAND             8122
BARCODE         841525
dtype: int64
The summary statistics of the dataset are:                CATEGORY_1 CATEGORY_2        CATEGORY_3 CATEGORY_4  \
count              845441     844128            784986      67459   
unique                 27        121               344        127   
top     Health & Wellness      Candy  Confection Candy  Lip Balms   
freq               512695     1210

In [4]:
# Look at the duplicated rows in the products dataset

duplicate_rows = products[products.duplicated()]
duplicate_rows.head(25)

CATEGORY_1              CATEGORY_2  \
91710   Health & Wellness               Skin Care   
109639         Restaurant               Beverages   
126613         Restaurant               Beverages   
128662             Snacks             Snack Cakes   
136757         Restaurant               Beverages   
140419         Restaurant               Beverages   
183128         Restaurant               Beverages   
205219         Restaurant               Beverages   
210811             Snacks             Snack Cakes   
225796         Restaurant               Beverages   
264937         Restaurant               Beverages   
293036         Restaurant               Beverages   
306395         Restaurant               Beverages   
326114         Restaurant               Beverages   
349073         Restaurant               Beverages   
349341             Snacks                   Chips   
359328            Alcohol                    Beer   
370474  Health & Wellness                     NaN   
371230         Restaurant               Beverages   
379611         Restaurant               Beverages   
400431             Snacks             Snack Cakes   
408671         Restaurant               Beverages   
410161  Health & Wellness  Medicines & Treatments   
438061             Snacks                   Candy   
442467         Restaurant               Beverages   

                                    CATEGORY_3      CATEGORY_4  \
91710                                      NaN             NaN   
109639                                    Soda       Diet Soda   
126613                                    Soda             NaN   
128662                     Brownie Snack Cakes             NaN   
136757                                    Soda             NaN   
140419                                    Soda             NaN   
183128                                    Soda             NaN   
205219                                    Soda             NaN   
210811            Danishes & Puffs Snack Cakes             NaN   
225796                                    Soda             NaN   
264937                                    Soda             NaN   
293036                                    Soda             NaN   
306395                        Slushies & Icees             NaN   
326114                                    Soda             NaN   
349073                                    Soda             NaN   
349341                                  Crisps             NaN   
359328                                   Lager  American Lager   
370474                                     NaN             NaN   
371230                                    Soda             NaN   
379611                                    Soda       Diet Soda   
400431                     Brownie Snack Cakes             NaN   
408671                                    Soda             NaN   
410161  Allergy & Sinus Medicines & Treatments             NaN   
438061                        Confection Candy             NaN   
442467                                    Soda             NaN   

                  MANUFACTURER                          BRAND BARCODE  
91710         BEIERSDORF, INC.                     COPPERTONE     NaN  
109639                 PEPSICO                          PEPSI     NaN  
126613                 PEPSICO                          PEPSI     NaN  
128662                   BIMBO  ENTENMANN'S SWEET BAKED GOODS     NaN  
136757   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
140419   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
183128   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
205219   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
210811                   BIMBO  ENTENMANN'S SWEET BAKED GOODS     NaN  
225796   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
264937   THE COCA-COLA COMPANY                      COCA-COLA     NaN  
293036   THE COCA-COLA COMPANY                      COC

### Observations 

* The `products` table includes 4 category/taxonomy levels where `CATEGORY_1` appears to be the parent to the child categories (levels 2-4) depending on taxonomy depth that the product falls under.
  * a NULL value at `CATEGORY_3` or `CATEGORY_4` indicates no further Fetch category for the product to fall under
* Manufacturer and brand can go hand-in-hand where several brands could roll up to a manufacturer that creates the products/brand/own the IP. This could provide opportunity to view transaction or user insights by not only individual brand, but rolling up to manufacturer.
  * A NULL value in either or both `MANUFACTURER` or `BRAND` column indicates there is not enough information from Fetch, the receipt or the CPG to identify brand. Additional modeling or analysis would be required to tie the barcode (assuming barcodes don't tie to other product/product sizes) to the appropriate product to determine manufacturer and brand.
* Barcode (likely also UPC) can have leading or trailing zeroes, will need to further investigate or keep in mind that barcode could be shared within different retailers/partners/manufacturers. 
* Duplicate rows indicate taxonomy, manufacturer, brand data with no barcode attached. Given this is in the products table, these rows should be removed as there is no primary key (PK) to tie them to a respective product in this table or another provided table. 
   * If I wanted to dig deeper, I could determine how many of these exact duplicate rows matching by taxonomy and manufacturer/brand pair to products with a barcode to determine if there could be a barcode assigned to the missing product.

In [5]:
transactions.head()

RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
4  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

        BARCODE FINAL_QUANTITY FINAL_SALE  
0  015300014978           1.00             
1           NaN           zero       1.49  
2  078742229751           1.00             
3  783399746536           zero       3.49  
4  047900501183           1.00

In [6]:
data_quality_check(transactions)

The shape of the dataset is: (50000, 8)
The columns in the dataset are: Index(['RECEIPT_ID', 'PURCHASE_DATE', 'SCAN_DATE', 'STORE_NAME', 'USER_ID',
       'BARCODE', 'FINAL_QUANTITY', 'FINAL_SALE'],
      dtype='object')
The number of missing values in the dataset are: RECEIPT_ID           0
PURCHASE_DATE        0
SCAN_DATE            0
STORE_NAME           0
USER_ID              0
BARCODE           5762
FINAL_QUANTITY       0
FINAL_SALE           0
dtype: int64
The number of unique values in the dataset are: RECEIPT_ID        24440
PURCHASE_DATE        89
SCAN_DATE         24440
STORE_NAME          954
USER_ID           17694
BARCODE           11028
FINAL_QUANTITY       87
FINAL_SALE         1435
dtype: int64
The summary statistics of the dataset are:                                   RECEIPT_ID PURCHASE_DATE  \
count                                  50000         50000   
unique                                 24440            89   
top     bedac253-2256-461b-96af-267748e6cecf    202

In [7]:
duplicated_transactions = transactions[transactions.duplicated()].sort_values(by='RECEIPT_ID')

SQL = """

SELECT * 
FROM transactions
WHERE RECEIPT_ID IN (SELECT RECEIPT_ID
                     FROM transactions
                     GROUP BY RECEIPT_ID
                     HAVING COUNT(*) > 1)
ORDER BY RECEIPT_ID
;
"""

# Examples for observation notes
sampled_duplicate_transactions = transactions[transactions['RECEIPT_ID'].isin(['4ec870d2-c39f-4a40-bf8a-26a079409b20', '1c12bd8a-c68c-41ee-a26d-294021d3e0b8'])].sort_values(by='RECEIPT_ID')
sampled_duplicate_transactions


RECEIPT_ID PURCHASE_DATE  \
2722   1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
2723   1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
2724   1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
45038  1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
45039  1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
45040  1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
7673   4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
7674   4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
7675   4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
7676   4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
48294  4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
48295  4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
48296  4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   
48311  4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   

                       SCAN_DATE STORE_NAME                   USER_ID  \
2722   2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
2723   2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
2724   2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
45038  2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
45039  2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
45040  2024-09-07 15:44:35.241 Z        CVS  65af09757050d0a6206ab136   
7673   2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
7674   2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
7675   2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
7676   2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
48294  2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
48295  2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
48296  2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   
48311  2024-09-08 19:39:01.589 Z    WALMART  638e9ae602a4e512e0585b59   

            BARCODE FINAL_QUANTITY FINAL_SALE  
2722            NaN           1.00             
2723            NaN           zero       0.00  
2724            NaN           1.00             
45038           NaN           1.00       0.00  
45039           NaN           1.00       0.00  
45040           NaN           1.00       0.00  
7673   072392016358           zero       1.48  
7674   072392016358           1.00             
7675   072392016358           zero       1.48  
7676   072392016358           1.00             
48294  072392016358           1.00       1.48  
48295  072392016358           1.00       1.48  
48296  072392016358           1.00       1.48  
48311  072392016358           1.00       1.48

### Observations

* The `transactions` table provides an unique receipt_id for all the respective barcodes purchased on the date, scan date (scanned by user into Fetch app (?)), store purchaesd from, and the user_id the transaction belongs to. `FINAL_QUANTITY` and `FINAL_SALE` indicates that an unique barcode to the receipt_id should total the quantity of products per barcode with the sale price for the product; this logic should ensure there are no duplicate barcodes per receipt_id.
* Scan date does not indicate timezone or UTC 
* Not all duplicate rows have the same case (where the duplicates in the `products` table lacked a barcode identifier)
  * Examples of cases with duplicates in `transactions`:
    * Receipt_id `4ec870d2-c39f-4a40-bf8a-26a079409b20` has 5 duplicate rows for the same transaction with the same scan_date and barcode, but is missing an entry for `FINAL_QUANTITY` and another row for `FINAL SALE`
    * Receipt_id `1c12bd8a-c68c-41ee-a26d-294021d3e0b8` has 3 duplicate rows but is missing a barcode for all three rows, so it is unclear if barcodes are missing for the same product, or three products from the one transaction. The `FINAL_QUANTITY` and `FINAL_SALE` columns do not provide enough transaction data to make an assumption if they could be all different or the same product.
  * For the sake of a quick analysis, these should be removed from the dataset, but would need to follow up with Data Engineering/Analytics Engineering counterparts to determine root cause issue of duplicate entries in the table.

In [8]:
users.head()

ID               CREATED_DATE  \
0  5ef3b4f17053ab141787697d  2020-06-24 20:17:54.000 Z   
1  5ff220d383fcfc12622b96bc  2021-01-03 19:53:55.000 Z   
2  6477950aa55bb77a0e27ee10  2023-05-31 18:42:18.000 Z   
3  658a306e99b40f103b63ccf8  2023-12-26 01:46:22.000 Z   
4  653cf5d6a225ea102b7ecdc2  2023-10-28 11:51:50.000 Z   

                  BIRTH_DATE STATE LANGUAGE  GENDER  
0  2000-08-11 00:00:00.000 Z    CA   es-419  female  
1  2001-09-24 04:00:00.000 Z    PA       en  female  
2  1994-10-28 00:00:00.000 Z    FL   es-419  female  
3                        NaN    NC       en     NaN  
4  1972-03-19 00:00:00.000 Z    PA       en  female

In [9]:
data_quality_check(users)

The shape of the dataset is: (100000, 6)
The columns in the dataset are: Index(['ID', 'CREATED_DATE', 'BIRTH_DATE', 'STATE', 'LANGUAGE', 'GENDER'], dtype='object')
The number of missing values in the dataset are: ID                  0
CREATED_DATE        0
BIRTH_DATE       3675
STATE            4812
LANGUAGE        30508
GENDER           5892
dtype: int64
The number of unique values in the dataset are: ID              100000
CREATED_DATE     99942
BIRTH_DATE       54721
STATE               52
LANGUAGE             2
GENDER              11
dtype: int64
The summary statistics of the dataset are:                               ID               CREATED_DATE  \
count                     100000                     100000   
unique                    100000                      99942   
top     5ef3b4f17053ab141787697d  2023-01-12 18:30:15.000 Z   
freq                           1                          2   

                       BIRTH_DATE  STATE LANGUAGE  GENDER  
count                   

### Observations

* The `users` table is simple and clean. There is an unique `ID` for the respective user, when they created an account at Fetch, followed by 4 fields (birthdate, state, language, gender) that are likely self-reported by the user including data that can be used for insights

In [10]:
# Remove duplicate rows from all datasets for Part 2 analysis

products.drop_duplicates(inplace=True)
transactions.drop_duplicates(inplace=True)

# If a product from a `transactions` did not have a FINAL_QUANTITY, fill with a float rather than a string

transactions['FINAL_QUANTITY'] = transactions['FINAL_QUANTITY'].replace('zero', 0).astype(float)

### Second: Provide SQL Queries

**Answer three of the following questions with at least one question coming from the closed-ended and one from the open-ended question set. Each question should be answered using one query.**

**Closed-ended questions:**

- What are the top 5 brands by receipts scanned among users 21 and over?
- What are the top 5 brands by sales among users that have had their account for at least six months?
- What is the percentage of sales in the Health & Wellness category by generation?

**Open-ended questions:** *(For these, make assumptions and clearly state them when answering the question.)*

- Who are Fetch’s power users?
- Which is the leading brand in the Dips & Salsa category?
- At what percent has Fetch grown year over year?

In [ ]:
#--------------------------------------------------------
# Top 5 brands by receipts scanned among users 21+
#--------------------------------------------------------

# Change BIRTH_DATE to datetime format
users['BIRTH_DATE'] = pd.to_datetime(users['BIRTH_DATE']).dt.strftime('%Y-%m-%d')

SQL = """
SELECT 
  p.BRAND_NAME, 
  COUNT(t.RECEIPT_ID) AS receipts_scanned
FROM transactions t
JOIN products p
  ON t.BARCODE = p.BARCODE
JOIN users u
  ON t.USER_ID = u.ID
WHERE DATE_DIFF(CURRENT_DATE(), DATE(u.BIRTH_DATE), YEAR) >= 21
GROUP BY p.BRAND_NAME
ORDER BY receipts_scanned DESC
LIMIT 5
;
"""

# Filter users who are 21+
users['BIRTH_DATE'] = pd.to_datetime(users['BIRTH_DATE'])
users_21_plus = users[users['BIRTH_DATE'] <= pd.Timestamp('today') - pd.DateOffset(years=21)]

# Merge transactions with products and users
merged_df = transactions.merge(products, on='BARCODE').merge(users_21_plus, left_on='USER_ID', right_on='ID')

# Group by brand and count the number of receipts scanned
top_brands = merged_df.groupby('BRAND')['RECEIPT_ID'].count().reset_index()

# Sort and get the top 5 brands
top_5_brands = top_brands.sort_values(by='RECEIPT_ID', ascending=False).head(5)

top_5_brands


BRAND  RECEIPT_ID
37   ANNIE'S HOMEGROWN GROCERY         552
204                       DOVE         535
69                    BAREFOOT         529
535                      ORIBE         483
57                      AVEENO         460

## **Top 5 Brands by Receipts Scanned (Users 21 and Over)**  

Based on the available data, the top 5 brands with the highest number of receipts scanned among users **21+** are:  

| **Brand**                    | **Receipts Scanned** |
|------------------------------|----------------------|
| ANNIE'S HOMEGROWN GROCERY    | 552                  |
| DOVE                         | 535                  |
| BAREFOOT                     | 529                  |
| ORIBE                        | 483                  |
| AVEENO                       | 460                  |


In [ ]:
#------------------------------------------------------------
# Top 5 Brands by Sales for Habitual Users (> 6 mo acct age)
#------------------------------------------------------------

SQL = """
SELECT 
    p.BRAND_NAME,
    SUM(COALESCE(t.FINAL_SALE, 0) * COALESCE(t.FINAL_QUANTITY, 0)) AS total_sales
FROM transactions t
JOIN products p 
  ON t.BARCODE = p.BARCODE
JOIN users u 
  ON t.USER_ID = u.ID
WHERE u.CREATED_DATE <= DATE_ADD(CURRENT_DATE, INTERVAL -6 MONTH)
GROUP BY p.BRAND_NAME
ORDER BY total_sales DESC
LIMIT 5
;
""" 

# Convert CREATED_DATE to datetime and assume timezone-aware
users['CREATED_DATE'] = pd.to_datetime(users['CREATED_DATE'], utc=True)

# Filter users who have had their account for at least 6 months
six_months_ago = pd.Timestamp.now(tz='UTC') - pd.DateOffset(months=6)
users_6_months = users[users['CREATED_DATE'] <= six_months_ago]

# Merge transactions with products and users
merged_df_6_months = transactions.merge(products, on='BARCODE') \
                                 .merge(users_6_months, left_on='USER_ID', right_on='ID')

# Convert FINAL_SALE and FINAL_QUANTITY to numeric 
# pandas stored them at dtype object because of the 'zero' string
merged_df_6_months['FINAL_SALE'] = pd.to_numeric(merged_df_6_months['FINAL_SALE'], errors='coerce').fillna(0)
merged_df_6_months['FINAL_QUANTITY'] = pd.to_numeric(merged_df_6_months['FINAL_QUANTITY'], errors='coerce').fillna(0)

# Calculate TOTAL SALES (TS = FINAL_SALE * FINAL_QUANTITY)
merged_df_6_months['TOTAL_SALES'] = merged_df_6_months['FINAL_SALE'] * merged_df_6_months['FINAL_QUANTITY']

# Group by brand and sum the total sales
top_brands_sales = merged_df_6_months.groupby('BRAND')['TOTAL_SALES'].sum().reset_index()
# Sort and get the top 5 brands by total sales
top_5_brands_sales = top_brands_sales.sort_values(by='TOTAL_SALES', ascending=False).head(5)

top_5_brands_sales



BRAND  TOTAL_SALES
37   ANNIE'S HOMEGROWN GROCERY      1481.28
204                       DOVE      1450.47
69                    BAREFOOT      1419.56
535                      ORIBE      1296.12
57                      AVEENO      1234.40

## **Top 5 Brands by Sales (Users with Accounts ≥ 6 Months)**  

For users who have had their account for at least **six months**, the top 5 brands by total sales are:  

| **Brand**                    | **Total Sales ($)** |
|------------------------------|--------------------|
| ANNIE'S HOMEGROWN GROCERY    | 1481.28           |
| DOVE                         | 1450.47           |
| BAREFOOT                     | 1419.56           |
| ORIBE                        | 1296.12           |
| AVEENO                       | 1234.40           |


In [ ]:
#---------------------------
# Fetch Growth YoY (%)
#---------------------------

# Dataset runs through 2024-06-12 to 2024-09-08
# Not enough data to calculate YoY growth, but can calculate bi-weekly and MoM growth
# This will still help understand the growth rate over this quarter

# --------------------- Bi-Weekly Growth ---------------------

SQL = """
WITH biweekly_sales AS (
    SELECT 
        FLOOR(DATEDIFF(t.PURCHASE_DATE, (SELECT MIN(PURCHASE_DATE) FROM transactions)) / 14) AS biweek,
        MIN(t.PURCHASE_DATE) AS start_date,
        MAX(t.PURCHASE_DATE) AS end_date,
        SUM(COALESCE(t.FINAL_SALE, 0) * COALESCE(t.FINAL_QUANTITY, 0)) AS total_sales
    FROM transactions t
    GROUP BY biweek
)
SELECT 
    biweek,
    start_date,
    end_date,
    total_sales,
    LAG(total_sales) OVER (ORDER BY biweek) AS previous_biweek_sales,
    ROUND(
        100.0 * (total_sales - LAG(total_sales) OVER (ORDER BY biweek)) / NULLIF(LAG(total_sales) OVER (ORDER BY biweek), 0),
        2
    ) AS biweekly_growth_percent
FROM biweekly_sales
ORDER BY biweek
;
"""

# Convert PURCHASE_DATE to datetime
transactions['PURCHASE_DATE'] = pd.to_datetime(transactions['PURCHASE_DATE'])

# Convert FINAL_SALE and FINAL_QUANTITY to numeric
transactions['FINAL_SALE'] = pd.to_numeric(transactions['FINAL_SALE'], errors='coerce').fillna(0)
transactions['FINAL_QUANTITY'] = pd.to_numeric(transactions['FINAL_QUANTITY'], errors='coerce').fillna(0)

# Compute total sales
transactions['TOTAL_SALES'] = transactions['FINAL_SALE'] * transactions['FINAL_QUANTITY']

# --------------------- Biweekly Growth ---------------------

transactions['BIWEEK'] = ((transactions['PURCHASE_DATE'] - transactions['PURCHASE_DATE'].min()).dt.days // 14)

# Total sales per bi-week
biweekly_sales = transactions.groupby('BIWEEK').agg(
    start_date=('PURCHASE_DATE', 'min'),
    end_date=('PURCHASE_DATE', 'max'),
    total_sales=('TOTAL_SALES', 'sum')
).reset_index()

# Calculate biweekly growth
biweekly_sales['PREVIOUS_BIWEEK_SALES'] = biweekly_sales['total_sales'].shift(1)
biweekly_sales['BIWEEKLY_GROWTH_PERCENT'] = 100 * (
    (biweekly_sales['total_sales'] - biweekly_sales['PREVIOUS_BIWEEK_SALES']) / biweekly_sales['PREVIOUS_BIWEEK_SALES']
)

biweekly_sales


BIWEEK start_date   end_date  total_sales  PREVIOUS_BIWEEK_SALES  \
0       0 2024-06-12 2024-06-25   19123.3316                    NaN   
1       1 2024-06-26 2024-07-09   23101.5492             19123.3316   
2       2 2024-07-10 2024-07-23   22143.3320             23101.5492   
3       3 2024-07-24 2024-08-06   19901.4702             22143.3320   
4       4 2024-08-07 2024-08-20   21623.4919             19901.4702   
5       5 2024-08-21 2024-09-03   20005.9609             21623.4919   
6       6 2024-09-04 2024-09-08    5847.4727             20005.9609   

   BIWEEKLY_GROWTH_PERCENT  
0                      NaN  
1                20.802953  
2                -4.147848  
3               -10.124320  
4                 8.652736  
5                -7.480434  
6               -70.771348

In [ ]:
# --------------------- Month-over-Month Growth ---------------------

SQL = """
WITH monthly_sales AS (
    SELECT 
        DATE_FORMAT(t.PURCHASE_DATE, '%Y-%m') AS month,
        SUM(COALESCE(t.FINAL_SALE, 0) * COALESCE(t.FINAL_QUANTITY, 0)) AS total_sales
    FROM transactions t
    GROUP BY month
)
SELECT 
    month,
    total_sales,
    LAG(total_sales) OVER (ORDER BY month) AS previous_month_sales,
    ROUND(
        100.0 * (total_sales - LAG(total_sales) OVER (ORDER BY month)) / NULLIF(LAG(total_sales) OVER (ORDER BY month), 0),
        2
    ) AS mom_growth_percent
FROM monthly_sales
ORDER BY month
;
"""

# Extract month (YYYY-MM)
transactions['MONTH'] = transactions['PURCHASE_DATE'].dt.to_period('M')

# Total sales per month
monthly_sales = transactions.groupby('MONTH')['TOTAL_SALES'].sum().reset_index()

# Calculate MoM growth
monthly_sales['PREVIOUS_MONTH_SALES'] = monthly_sales['TOTAL_SALES'].shift(1)
monthly_sales['MOM_GROWTH_PERCENT'] = 100 * (
    (monthly_sales['TOTAL_SALES'] - monthly_sales['PREVIOUS_MONTH_SALES']) / monthly_sales['PREVIOUS_MONTH_SALES']
)

monthly_sales


MONTH  TOTAL_SALES  PREVIOUS_MONTH_SALES  MOM_GROWTH_PERCENT
0  2024-06   26614.3073                   NaN                 NaN
1  2024-07   49219.3168            26614.3073           84.935555
2  2024-08   46280.1917            49219.3168           -5.971487
3  2024-09    9632.7927            46280.1917          -79.185927

## **Estimating Fetch’s Year-over-Year (YoY) Growth**  

There isn't a full year of data, so instead, I will use biweekly trends and MoM growth to supplement. 
- **July sees +85% MoM growth**, likely from increased engagement or campaigns.  
- **August dipped slightly (-6%)**, while we want to see positive growth, this could suggest normalization. Consider stronger BTS CPG campaigns to drive engagement.  
- **September observes -79% drop** due to incomplete data for the month, not an actual business decline.  

### Key Assumptions 
- Data is valid — low risk of duplicates or errors.  
- September’s drastic decline is due to missing/incomplete data, not performance.  
- If this were a real business case, I'd recommending syncing with Analytics Engineering to validate accuracy for future consistency.  

True YoY growth can't be calculated without a full year of data, but the early data suggests Fetch had strong summer before hitting some expected fluctuations. I would recommend having a DS team investigate seasonality to determine future assumptions and expected fluctuations.

### Third: Communicate with Stakeholders

**Construct an email or Slack message that is understandable to a product or business leader who is not familiar with your day-to-day work. Summarize the results of your investigation. Include:**

- Key data quality issues and outstanding questions about the data
- One interesting trend in the data
  - Use a finding from part 2 or come up with a new insight
- Request for action: explain what additional help, info, etc. you need to make sense of the data and resolve any outstanding issues


Hi [Stakeholder/Product/Business Leader]!!! :cute-fun-smiley-slackmoji-here:  

I just finished my analysis on last year's summer transactions - here is a brief summary of my observations, key questions answered, and some next steps based on the trends I found. My SQL queries and additional details are documented here. **[Fetch Data Analysis](https://github.com/sabrinanoelle/fetch-25/blob/main/data-analyst-take-home.ipynb)**. 

**Observations & Data Quality Notes**  
- **September’s -79% drop in sales** is likely due to **incomplete data**, not a true business decline. Would love confirmation on whether this is expected, or if we need to pull additional data to see the impact at the start of the fall season.  
- **Receipt accuracy -** Assuming the dataset is clean, but are there any known issues with duplicate transactions or misscanned receipts?  
- Analysis filtered for accounts **6+ months old**, but should we refine this further to remove inactive users?  
- Housekeeping/Macrodata Refinement: Data looks well-structured, but if this were a real case, I’d want to verify low risk of errors with Analytics Engineering.  

**Answers to requested questions**  

**Top 5 Brands by Receipts Scanned (Users 21+)**  
| **Brand**                    | **Receipts Scanned** |
|------------------------------|----------------------|
| ANNIE'S HOMEGROWN GROCERY    | 552                  |
| DOVE                         | 535                  |
| BAREFOOT                     | 529                  |
| ORIBE                        | 483                  |
| AVEENO                       | 460                  |

**TLDR:** These brands see **high engagement**, likely due to strong consumer preference or frequent purchase behavior.  

**Top 5 Brands by Sales (Users with Accounts ≥ 6 Months)**  
| **Brand**                    | **Total Sales ($)** |
|------------------------------|--------------------|
| ANNIE'S HOMEGROWN GROCERY    | 1481           |
| DOVE                         | 1450           |
| BAREFOOT                     | 1420           |
| ORIBE                        | 1296           |
| AVEENO                       | 1234           |

**TLDR:** Long-term users show strong spending trends for these brands, which could indicate **brand loyalty or repeat purchases**.  

**3Growth Trends (Biweekly & MoM)**  
| MONTH  | TOTAL_SALES | MOM_GROWTH_PERCENT |
|--------|------------|--------------------|
| 2024-06 | 26,614.31  | NaN                |
| 2024-07 | 49,219.32  | **+85%**        |
| 2024-08 | 46,280.19  | **-6%**         |
| 2024-09 | 9,632.79   | **-79%**        |

**July was a massive growth month (+85% MoM)**; possible marketing push or seasonal trend.  
**August stabilized (-6%)**, suggesting normal user behavior.  
**September dropped (-79%),** likely due to incomplete data rather than an actual crash in sales.  

Could you please clear up some additional thoughts and questions I had for this analysis?: 
* Can we confirm if September’s data is fully available? This would validate whether the drop is real.  
* Were there any promotions, user engagement strategies, or external factors in July-August that explain the spike?  
* A quick sync with Analytics Engineering would help confirm if there are any data inconsistencies (e.g., duplicates, scanning issues).  

Let me know if you’d like a deeper walkthrough of my findings — I’m happy to refine further based on business needs. Thank you!  